### ISSUE TITLE
Portal domain routing with Docker

### ISSUE THREAD

**Issue by byrnedo** [TykTechnologies/tyk#225]
Hi, 
I'm running the gateway and dashboard on docker (not using the quickstart though) and everything works beautifully. 
I'm using consul for dns and discovery and tyk is sitting behind an nginx proxy which is templated by consul.

I'm getting a problem with setting up the portal though. 

1. As I understand, the portal for an organisation is available at `<domain>/<org-id>/portal`, but do I have to manually proxy `<org-slug>.<some-domain>` to this? 

2. I see that there's the `"portal_domains"` option in the config, but since that's hard coded, must I just manually edit that every time I want to add an org? 

3. Would the expected syntax for `portal_domains` be 
```
    "portal_domains": {
        "<org-id>": "my.lovely.domain"
    }
```

EDIT: I'm seeing that you proxy portal urls in the docker quickstart. Is that the best way to do this?
?

**Comment by lonelycode** [comment]
We only proxy the portal in the quickstart because it's single-instance and people like everything running on port 80.

You can also set the portal domain as a CNAME entry in the organisation record in MongoDB. you can use the organisations admin API to set this value (cname and cname_enabled).

The simplest method is to set the domain by using the dashboard, selecting "Your developer portal" and then "Set domain", this does the same thing as the above API call, just in a more limited context.

**Comment by byrnedo** [comment]
I had tried changing the cname in the gui, but I keep getting "Domain name is invalid, please enter a valid domain name" in the response. The domain I tried is "portal.local.sendify.se".

I think I tried adding it in mongo too but it didn't seem to work. I'll have to try that again.

**Comment by lonelycode** [comment]
You will need to do some domain binding for all of the services and make sure that the Tyk dashboard is configured to allow domains.

The details for all the settings are here:

https://tyk.io/v1.9/configuration/dashboard-config/#enable-host-names:8f1ac6c580c2f6cd2c8954a33f3ca4d6

You need to enable domains, and set a hostname for both the dashboard and the portal, then that can be overridden with the per org version.

Domains are port-bound, so if you are running on port 3000 then the domain that will work (and not 404) will need to include the port.

**Comment by byrnedo** [comment]
It's running on 3000 in the container. When you say "Domains are port bound" do you mean that the value for `hostname` in the conf would need to be "<somedomain>:3000" int that case?

**Comment by lonelycode** [comment]
No, sorry didn't mean to confuse, if you set a standard hostname Tyk will listen on hostname:port automatically depending on the port you set in the configuration.

So if it's on 3000, then setting a hostname of www.mydomain.com will only work on www.mydonain.com:3000 and that includes the portal.

Also pay attention to the portal root setting, everything will 404 except the portal root. 

The best thing to do might be to look at how the QuickStart handles things, although it's a bit special in that it reverse proxies the portal via The gateway so everything can be available on port 80.

However, if you are running behind NGinX, why bother with Tyk domain handling? Switch it off and have NGinX do the rewrites for you.

Take a look in the Tyk-analytics installation folder in the container, there is a host manager folder which has a bunch of templates for various Tyk routing configurations that demonstrate how the rewrites work for single site and subdomain based setups.

the host manager is a legacy service we used to bundle to manage domain routing so it's probably your best source of info for NGinX.

I think having Tyk try to bind to a domain *and* nginx do it you'll just end up confusing things. It might work, but it's two things to synchronise.

**Comment by byrnedo** [comment]
I found this thread on the community portal which is really the same thing
https://community.tyk.io/t/portal-and-subdomain/530/10.
Sorry for the duplicate.
I've settled for using tyk to proxy the resources for the portal. I have a
very convenient setup with nginx just now and am a little too lazy to mod
it to make room for the templating needed for proxying directly from it
based on org id. Or maybe I will, have a bit of time this weekend :)
